### ISSUE TITLE
ArrayIndexOutOfBoundsException in BlameAnnotationImpl.getLineRevisionNumber (191.0.7)

### ISSUE THREAD

**Issue by aaronols** [zielu/GitToolBox#135]
After updating to Version 191.0.7 of the Plugin I started to sometimes (note: not always) be notified on an ArrayIndexOutOfBoundsException in the Plugin.

Despite the exceptions the inline blame functionality continued to function properly.

There are two ways I managed to reproduce the issue:

1. Open a commited but unmodified file, copy a method and paste it below (Mark method, Ctrl+C, Esc, Enter, Enter, Ctrl+V). The error appears to trigger more often (but still not always) when doing this sequence faster. Update: The exception triggers when IntelliJ applies auto-formatting at the same time as new code is added.
The exception always occurs with value 134. Line number 134 is where the `return` statement of the pasted method is located after the paste. Before pasting the method, the file has 126 lines.
See the stacktrace below:

```
java.lang.ArrayIndexOutOfBoundsException: 134
	at zielu.gittoolbox.blame.BlameAnnotationImpl.getLineRevisionNumber(BlameAnnotationImpl.java:44)
	at zielu.gittoolbox.blame.BlameAnnotationImpl.getBlame(BlameAnnotationImpl.java:31)
	at zielu.gittoolbox.blame.BlameServiceImpl.getLineBlameInternal(BlameServiceImpl.java:107)
	at zielu.gittoolbox.blame.BlameServiceImpl.getLineBlameInternal(BlameServiceImpl.java:97)
	at zielu.gittoolbox.blame.BlameServiceImpl.lambda$getDocumentLineBlame$1(BlameServiceImpl.java:84)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.blame.BlameServiceImpl.getDocumentLineBlame(BlameServiceImpl.java:84)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineBlame(BlameEditorServiceImpl.java:147)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getInfosWithCaching(BlameEditorServiceImpl.java:110)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.lambda$getLineAnnotation$1(BlameEditorServiceImpl.java:84)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineAnnotation(BlameEditorServiceImpl.java:83)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.lambda$getLineExtensions$0(BlameEditorServiceImpl.java:73)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineExtensions(BlameEditorServiceImpl.java:73)
	at zielu.gittoolbox.ui.blame.BlameEditorLinePainter.getLineExtensions(BlameEditorLinePainter.java:19)
	at com.intellij.openapi.editor.impl.EditorImpl.processLineExtensions(EditorImpl.java:764)
	at com.intellij.openapi.editor.impl.view.EditorPainter.collectExtensions(EditorPainter.java:672)
	at com.intellij.openapi.editor.impl.view.EditorPainter.access$800(EditorPainter.java:49)
	at com.intellij.openapi.editor.impl.view.EditorPainter$1.paintAfterLineEnd(EditorPainter.java:256)
	at com.intellij.openapi.editor.impl.view.EditorPainter.paintLineFragments(EditorPainter.java:1263)
	at com.intellij.openapi.editor.impl.view.EditorPainter.paintBackground(EditorPainter.java:227)
	at com.intellij.openapi.editor.impl.view.EditorPainter.paint(EditorPainter.java:97)
	at com.intellij.openapi.editor.impl.view.EditorView.paint(EditorView.java:264)
	at com.intellij.openapi.editor.impl.EditorImpl.paint(EditorImpl.java:1914)
	at com.intellij.openapi.editor.impl.EditorComponentImpl.paintComponent(EditorComponentImpl.java:240)
	at javax.swing.JComponent.paint(JComponent.java:1057)
```

The other way I have reproduced this error (only once) is by reverting the last change to a file so it was unmodified again by clicking on the modified line marker and selecting `Rollback lines`. This produced an ArrayIndexOutOfBoundsException for position `-1`:

```
java.lang.ArrayIndexOutOfBoundsException: -1
	at zielu.gittoolbox.blame.BlameAnnotationImpl.getLineRevisionNumber(BlameAnnotationImpl.java:44)
	at zielu.gittoolbox.blame.BlameAnnotationImpl.getBlame(BlameAnnotationImpl.java:31)
	at zielu.gittoolbox.blame.BlameServiceImpl.getLineBlameInternal(BlameServiceImpl.java:107)
	at zielu.gittoolbox.blame.BlameServiceImpl.getLineBlameInternal(BlameServiceImpl.java:97)
	at zielu.gittoolbox.blame.BlameServiceImpl.lambda$getDocumentLineBlame$1(BlameServiceImpl.java:84)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.blame.BlameServiceImpl.getDocumentLineBlame(BlameServiceImpl.java:84)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineBlame(BlameEditorServiceImpl.java:147)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getInfosWithCaching(BlameEditorServiceImpl.java:110)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.lambda$getLineAnnotation$1(BlameEditorServiceImpl.java:84)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineAnnotation(BlameEditorServiceImpl.java:83)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.lambda$getLineExtensions$0(BlameEditorServiceImpl.java:73)
	at com.codahale.metrics.Timer.timeSupplier(Timer.java:133)
	at zielu.gittoolbox.ui.blame.BlameEditorServiceImpl.getLineExtensions(BlameEditorServiceImpl.java:73)
	at zielu.gittoolbox.ui.blame.BlameEditorLinePainter.getLineExtensions(BlameEditorLinePainter.java:19)
	at com.intellij.openapi.editor.impl.EditorImpl.processLineExtensions(EditorImpl.java:764)
	at com.intellij.openapi.editor.impl.view.EditorSizeManager.validateMaxLineWithExtension(EditorSizeManager.java:299)
	at com.intellij.openapi.editor.impl.view.EditorSizeManager.getTextPreferredWidth(EditorSizeManager.java:278)
	at com.intellij.openapi.editor.impl.view.EditorSizeManager.getPreferredSize(EditorSizeManager.java:181)
	at com.intellij.openapi.editor.impl.view.EditorView.getPreferredSize(EditorView.java:276)
	at com.intellij.openapi.editor.impl.EditorImpl.getPreferredSize(EditorImpl.java:2129)
	at com.intellij.openapi.editor.impl.EditorImpl.validateSize(EditorImpl.java:1771)
	at com.intellij.openapi.editor.impl.EditorImpl.changedUpdate(EditorImpl.java:1662)
	at com.intellij.openapi.editor.impl.EditorImpl.access$11900(EditorImpl.java:123)
	at com.intellij.openapi.editor.impl.EditorImpl$EditorDocumentAdapter.documentChanged(EditorImpl.java:4659)
	at com.intellij.openapi.editor.impl.DocumentImpl.changedUpdate(DocumentImpl.java:891)
	at com.intellij.openapi.editor.impl.DocumentImpl.updateText(DocumentImpl.java:796)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:661)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:608)
	at com.intellij.diff.util.DiffUtil.replaceLines(DiffUtil.java:914)
	at com.intellij.diff.util.DiffUtil.applyModification(DiffUtil.java:947)
	at com.intellij.openapi.vcs.ex.DocumentTracker.partiallyApplyBlocks(DocumentTracker.kt:207)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase$runBulkRollback$1.invoke(LineStatusTrackerBase.kt:337)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase$runBulkRollback$1.invoke(LineStatusTrackerBase.kt:39)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase$updateDocument$2.run(LineStatusTrackerBase.kt:149)
	at com.intellij.diff.util.DiffUtil.lambda$null$15(DiffUtil.java:1373)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:220)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:178)
	at com.intellij.diff.util.DiffUtil.lambda$executeWriteCommand$16(DiffUtil.java:1368)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1054)
	at com.intellij.diff.util.DiffUtil.executeWriteCommand(DiffUtil.java:1368)
	at com.intellij.diff.util.DiffUtil.executeWriteCommand(DiffUtil.java:1384)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase.updateDocument(LineStatusTrackerBase.kt:149)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase.runBulkRollback(LineStatusTrackerBase.kt:336)
	at com.intellij.openapi.vcs.ex.LineStatusTrackerBase.rollbackChanges(LineStatusTrackerBase.kt:323)
	at com.intellij.openapi.vcs.ex.RollbackLineStatusAction.rollback(RollbackLineStatusAction.java:35)
	at com.intellij.openapi.vcs.ex.LocalLineStatusTracker$LocalLineStatusMarkerRenderer$RollbackLineStatusRangeAction.actionPerformed(LineStatusTracker.kt:128)
	at com.intellij.openapi.vcs.ex.LineStatusMarkerPopupRenderer$RangeMarkerAction.actionPerformed(LineStatusMarkerPopupRenderer.java:434)
	at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:260)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:277)
	at com.intellij.openapi.actionSystem.impl.ActionButton.actionPerformed(ActionButton.java:183)
	at com.intellij.openapi.actionSystem.impl.ActionButton.performAction(ActionButton.java:147)
	at com.intellij.openapi.actionSystem.impl.ActionButton.processMouseEvent(ActionButton.java:415)
	at java.awt.Component.processEvent(Component.java:6315)
```

Please note that the inline blame functionality continues to work as expected regardless of these exceptions.

**Comment by zielu** [comment]
Seems like that change from map to array for storage was not such a great idea :/

**Comment by aaronols** [comment]
I understand the array is instantiated in the constructor of `BlameAnnotationImpl`.
When is this object created and/or re-created?

I noticed that I have automatic formatting enabled in IntelliJ. The automatic formatting only triggers after the first modification to the file.
So when I insert the two newlines (as described above), wait for the automatic formatting of IntelliJ to complete and then paste the new method, the exception does not occur.

Sounds like some sort of race condition (and/or incorrect use/reuse of LineExtensions) between painting the changes to the view for the pasted code and the changes for the automatic formatting.

**Comment by zielu** [comment]
Fixed in 191.0.9, leaving open to add tests for this case

**Comment by aaronols** [comment]
That was fast. I wanted to say that I appreciate that :)
